modules/node/src/utils.ts

```diff
@@ -0,0 +1,94 @@
+import {
+  TransferWithExpiryStatus,
+  AppName,
+  TransferStatus,
+  TransferStatuses,
+  TransferWithExpiryStatuses,
+  HashLockTransferAppState,
+} from "@connext/types";
+import { AppInstance, AppType } from "./appInstance/appInstance.entity";
+import { bigNumberifyJson } from "@connext/utils";
+
+export function appStatusesToTransferStatus<T extends AppName>(
+  senderApp: AppInstance<T>,
+  receiverApp?: AppInstance<T>,
+): TransferStatus | undefined {
+  switch (senderApp.type) {
+    case AppType.PROPOSAL: {
+      return TransferStatuses.PENDING;
+    }
+    case AppType.UNINSTALLED: {
+      if (!receiverApp) {
+        // Sender app was uninstalled and receiver app rejected
+        return TransferStatuses.FAILED;
+      }
+      if (receiverApp.type !== AppType.UNINSTALLED) {
+        // receiver app exists, while sender app has been uninstalled
+        return TransferStatuses.FAILED;
+      }
+      // sender and receiver have uninstalled their apps
+      // FIXME: How will you be able to determine if this was a
+      // collaborative payment cancellation/failure instead of a
+      // success?
+      // cant check latest state preimage on app because there is no
+      // guarantee the app state in our db corresponds to the latest
+      // state of the app (updating state to include preimage does not
+      // happen under lock)
```

Comment:
You're right - there's no way to really have the status be conditional upon latest state unless that state is updated under the lock. Maybe we *do* have to save the latest state after all...

This shouldn't block merging this PR for now, though.